rfd/0053-passwordless-fido2.md

```diff
@@ -0,0 +1,494 @@
+            │                   │ <─────────────────────                                             │
+            │                   │                      │                                             │
+            │      *taps*       │                      │                                             │
+            │<──────────────────│                      │                                             │
+            │                   │                      │                                             │
+            │            signed challenge              │                                             │
+            │─────────────────────────────────────────>│                                             │
+            │                   │                      │                                             │
+            │                   │                      │ AuthenticateSSH(userHandle, signedChallenge)│
+            │                   │                      │ ────────────────────────────────────────────>
+            │                   │                      │                                             │
+            │                   │                      │               SSH credentials               │
+            │                   │                      │ <────────────────────────────────────────────
+            │                   │                      │                                             │
+            │                   │          ok          │                                             │
+            │                   │ <─────────────────────                                             │
+     ┌──────┴──────┐           user                  ┌─┴─┐                                       ┌───┴────┐
+     │authenticator│            ┌─┐                  │tsh│                                       │Teleport│
+     └─────────────┘            ║"│                  └───┘                                       └────────┘
+                                └┬┘
+                                ┌┼┐
+                                 │
+                                ┌┴┐
+```
+
+(As a side note, it's interesting to note the differences in interaction between
+Web and `tsh` and how it may provide clues as to how browsers work.)
+
+(1) is the "key selection" step explained above, requiring one touch\* from the
+user to choose the desired authenticator.
+
+(2) is the PIN-entering step. For biometric authenticators, we can "merge" steps
+(1), (2) and (3) by using [Credentials](
+https://github.com/keys-pub/go-libfido2/blob/9246c0bee502c5f2c2495557f07fa5f45da8cd1c/fido2.go#L770)
+as the method to choose the authenticator. This technique requires one user
+touch/verification and avoids the need for PINs with biometric authenticators.
+
+(3) is when we enumerate resident keys for the {authenticator, RPID} pair. We
+can skip user interaction if only a single key exists.
```

Comment:
I'm guessing there will be a unique RPID per teleport cluster? Would it ever make sense for a user to have multiple credentials on the same hardware key for Teleport? If not, is there any way for us to prevent them from doing that?

Reply:
Correct, we have a unique RPID per cluster. It's a similar situation to the old U2F AppID.

They may have credentials for multiple Teleport users (probably rare for the average user, but I do it all the time). It's fine if they do, WebAuthn foresees that, plus I don't see any problems in allowing it.

Reply:
I meant, can a single Teleport user have multiple credentials on the same hardware key? IE both credentials reference the same Teleport user in the backend. I could see someone configuring that by mistake (maybe while trying to add a second/backup key) and getting confused by seeing multiple logins with the same name. Probably not a big deal though, they could just delete duplicates

Reply:
Oh, got it. [We don't allow that for WebAuthn](https://github.com/gravitational/teleport/blob/bd01152781443609fad53ab52722c9c38785eb53/lib/auth/webauthn/register.go#L138-L153), so should be safe in this front.